File: Source/Core/Utility/NSFetchedResultsController+Utility.m
//
//  NSFetchedResultsController+Utility.m
//  Eunomia
//
//  Created by Ian on 5/24/14.
//

#import "NSFetchedResultsController+Utility.h"

@implementation NSFetchedResultsController (Eunomia_Utility)

+ (id)firstResultWithFetchRequest:(NSFetchRequest *)fetchRequest
             managedObjectContext:(NSManagedObjectContext *)context
               sectionNameKeyPath:(NSString *)sectionNameKeyPath
                        cacheName:(NSString *)name
                            error:(NSError **)error
{
    NSFetchedResultsController *fetchResultsController = [ [NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest
                                                                                              managedObjectContext:context
                                                                                                sectionNameKeyPath:sectionNameKeyPath
                                                                                                         cacheName:name];
    
    return [fetchResultsController firstResultOfPerformFetch:error];
}

- (id)firstResultOfPerformFetch:(NSError **)error
{
    id result;
    [self performFetch:error];
    
    if (self.fetchedObjects.count > 0)
    {
        result = [self.fetchedObjects firstObject];
    }
    
    return result;
}

@end

File: Source/Core/UtilityExtensions/CoreData/NSFetchedResultsController+Utility.swift
//
//  NSFetchedResultsController+Utility.swift
//  Eunomia
//
//  Created by Ian Grossberg on 7/29/15.
//  Copyright (c) 2015 Adorkable. All rights reserved.
//

import Foundation

import CoreData

// Currently bugged: https://bugs.swift.org/browse/SR-2708
//extension NSFetchedResultsController {
//    
//    convenience init(fetchObjectType : NSManagedObject.Type, managedObjectContext context : NSManagedObjectContext, sectionNameKeyPath : String? = nil, cacheName : String? = nil) {
//        
//        let useFetchRequest = fetchObjectType.guaranteeFetchRequest(nil)
//        self.init(fetchRequest: useFetchRequest as! NSFetchRequest<_>, managedObjectContext: context, sectionNameKeyPath: sectionNameKeyPath, cacheName: cacheName)
//    }
//    
//    public func firstResultOfPerformFetch() throws -> AnyObject? {
//        var result : AnyObject?
//        
//        try self.performFetch()
//        
//        if let fetchedObjects = self.fetchedObjects
//        {
//            result = fetchedObjects.first
//        }
//        
//        return result
//    }
//    
//    public class func firstResult(_ fetchRequest : NSFetchRequest<NSFetchRequestResult>, context : NSManagedObjectContext, sectionNameKeyPath : String? = nil, cacheName : String? = nil) throws -> AnyObject? {
//        
//        let fetchedResultsController = NSFetchedResultsController(fetchRequest: fetchRequest as! NSFetchRequest<NSFetchRequestResult>, managedObjectContext: context, sectionNameKeyPath: sectionNameKeyPath, cacheName: cacheName)
//        return try fetchedResultsController.firstResultOfPerformFetch()
//    }
//    
//    public var fetchedObjectsCount : Int {
//        return guarantee(self.fetchedObjects?.count, fallback: 0)
//    }
//    
//    public class func fetchedObjectsCount(_ fetchObjectType : NSManagedObject.Type, context : NSManagedObjectContext, sectionNameKeyPath : String? = nil, cacheName : String? = nil) throws -> Int {
//        
//        let fetchedResultsController = NSFetchedResultsController(fetchObjectType: fetchObjectType, managedObjectContext: context, sectionNameKeyPath: sectionNameKeyPath, cacheName: cacheName)
//        try fetchedResultsController.performFetch()
//        
//        return fetchedResultsController.fetchedObjectsCount
//    }
//}
